Product Subscriptions: Analytics

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11183
<!-- Id number of the GitHub issue this PR addresses. -->

This PR adds the following analytics:

1. `product_list_loaded` event now includes `is_eligible_for_subscriptions` property.
2. Expiration events:
  - `product_details_view_subscription_expiration_date_tapped` for simple subscription products
  - `product_variation_view_subscription_expiration_date_tapped`  for variation subscription products
3. Free trial setting events:
  - `product_details_view_subscription_free_trial_tapped` for simple subscription products
  - `product_variation_view_subscription_free_trial_tapped`  for variation subscription products
4. Done tapping events: 
  - `product_subscription_expiration_done_button_tapped` event when "Done" button on Expiration setting is tapped, with `has_changed_data` property
  - `product_subscription_free_trial_done_button_tapped` event when "Done" button on Free Trial setting is tapped, with `has_changed_data` property
 
## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Have a test site containing both simple and variation subscription products,
2. In the app, open products list screen. Once products are loaded, check for event `product_list_loaded` with `is_eligible_for_subscriptions` set as true,
3. Open a simple subscription product,
4. Open the "Free trial" setting. Check for event `product_details_view_subscription_free_trial_tapped`
5. Change some settings there and tap Done. Check for event `product_subscription_free_trial_done_button_tapped` with `has_changed_data` property set as true.
6. Open the "Expire after" setting. Check for event `product_subscription_free_trial_done_button_tapped`. 
7. Change some settings there and tap Done. Check for event `product_subscription_expiration_done_button_tapped` with `has_changed_data` property set as true.
8. Open a variation subscription product,
9. Open the "Free trial" setting. Check for event `product_variation_view_subscription_free_trial_tapped`
10. Change some settings there and tap Done. Check for event `product_subscription_free_trial_done_button_tapped` with `has_changed_data` property set as true.
11. Open the "Expire after" setting. Check for event `product_variation_view_subscription_expiration_date_tapped`. 
12. Change some settings there and tap Done. Check for event `product_subscription_expiration_done_button_tapped` with `has_changed_data` property set as true.

hafizrahman committed:
Add custom property `is_eligible_for_subscriptions` for productListLoaded.

hafizrahman committed:
Add Woo Subscriptions eligibility checker to ProductListViewModel.

hafizrahman committed:
Update productListLoaded tracking to include subscriptions eligibility property.

hafizrahman committed:
Merge branch 'trunk'

hafizrahman committed:
Add analytics for viewing Subscription expiration setting, both for simple and variation subscriptions product.

hafizrahman committed:
Add analytics for viewing Free trial setting, both for simple and variation subscriptions product.

hafizrahman committed:
Add tracking for when Done button is tapped in Subscription Expiration Details screen.

hafizrahman committed:
Add tracking for when Done button is tapped in Subscription Free Trial screen.

Also fixes some typos.

hafizrahman committed:
Remove trailing whitespace.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -516,13 +516,25 @@ final class ProductFormViewController<ViewModel: ProductFormViewModelProtocol>:
                 }
-                // TODO: 11090 - Analytics
+
+                if product is EditableProductVariationModel {
+                    ServiceLocator.analytics.track(.productVariationViewSubscriptionFreeTrialTapped)
+                } else if product is EditableProductModel {
+                    ServiceLocator.analytics.track(.productDetailsViewSubscriptionFreeTrialTapped)
+                }
```
Instead of checking the product model's type, we should be using `eventLogger` property from this view controller to log these events. [Example code](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Products/Edit%20Product/ProductFormViewController.swift#L397)

hafizrahman replied:
TIL about this. Added in https://github.com/woocommerce/woocommerce-ios/pull/11236/commits/f4fef50761ff162c8e512bf4d3532f467e598a0a

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -516,13 +516,25 @@ final class ProductFormViewController<ViewModel: ProductFormViewModelProtocol>:
                 }
-                // TODO: 11090 - Analytics
+
+                if product is EditableProductVariationModel {
+                    ServiceLocator.analytics.track(.productVariationViewSubscriptionExpirationDateTapped)
+                } else if product is EditableProductModel {
+                    ServiceLocator.analytics.track(.productDetailsViewSubscriptionExpirationDateTapped)
+                }
```
Same as the other comment. We should use the `eventLogger` to log these events.

hafizrahman replied:
Added in https://github.com/woocommerce/woocommerce-ios/pull/11236/commits/ad05c6db36ae416e9b1e941423d8dace72e31e10

selanthiraiyan reviewed (changes requested):
Thanks for working on this, Hafiz!

I added two comments about using `eventLogger` property to log events instead of type checking the product model to differentiate between product and variation form. 

Kindly let me know your thoughts.

hafizrahman committed:
Refactor subscription free trial row tapping event to use eventLogger.

hafizrahman committed:
Refactor subscription expiration date row tapping event to use eventLogger.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormEventLogger.swift:
```diff
@@ -31,4 +31,12 @@ struct ProductFormEventLogger: ProductFormEventLoggerProtocol {
     }
+
+    func logSubscriptionsFreeTrialTapped() {
+        ServiceLocator.analytics.track(.productDetailsViewSubscriptionFreeTrialTapped)
```
Nit: Consider following the pattern used in `func logSubscriptionsTapped()` above which calls `func subscriptionsTapped()` to get the event.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11236/commits/20e2f8a8d73d989bd6928e576cc3ace90b59754a

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormEventLogger.swift:
```diff
@@ -31,4 +31,12 @@ struct ProductFormEventLogger: ProductFormEventLoggerProtocol {
+    }
+
+    func logSubscriptionsExpirationDateTapped() {
+        ServiceLocator.analytics.track(.productDetailsViewSubscriptionExpirationDateTapped)
```
Nit: Consider following the pattern used in `func logSubscriptionsTapped()` above which calls `func subscriptionsTapped()` to get the event.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/commit/20e2f8a8d73d989bd6928e576cc3ace90b59754a

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductVariationFormEventLogger.swift:
```diff
@@ -31,4 +31,12 @@ struct ProductVariationFormEventLogger: ProductFormEventLoggerProtocol {
     }
+
+    func logSubscriptionsFreeTrialTapped() {
+        ServiceLocator.analytics.track(.productVariationViewSubscriptionFreeTrialTapped)
```
Same as the other comment about following the pattern from `func logSubscriptionsTapped()` for improved readability.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/commit/20e2f8a8d73d989bd6928e576cc3ace90b59754a

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductVariationFormEventLogger.swift:
```diff
@@ -31,4 +31,12 @@ struct ProductVariationFormEventLogger: ProductFormEventLoggerProtocol {
+    }
+
+    func logSubscriptionsExpirationDateTapped() {
+        ServiceLocator.analytics.track(.productVariationViewSubscriptionExpirationDateTapped)
```
Same as the other comment about following the pattern from `func logSubscriptionsTapped()` for improved readability.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/commit/20e2f8a8d73d989bd6928e576cc3ace90b59754a

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift:
```diff
@@ -11,13 +12,22 @@ class ProductListViewModel {
         self.siteID = siteID
         self.stores = stores
+        self.storage = storage
+        self.wooSubscriptionProductsEligibilityChecker = WooSubscriptionProductsEligibilityChecker(
+            siteID: siteID,
+            storage: storage
+        )
```
I see that you aren't using the `storage` to write unit tests for `ProductListViewModel`. 

- In this case you can skip injecting it into `ProductListViewModel` class.
- You can just pass the `siteID` to `WooSubscriptionProductsEligibilityChecker`'s init. 

```suggestion
        self.wooSubscriptionProductsEligibilityChecker = WooSubscriptionProductsEligibilityChecker(siteID: siteID)
```

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11236/commits/108f60ac9d727cc958f44efadd6810eb36800c91

(not using the suggestion above because I need to add a bit more code changes)

selanthiraiyan reviewed (approved):
Thanks for the updates, Hafiz! This tracks as expected. 🎉 

I left a few nits. :shipit:

hafizrahman committed:
Remove unneeded storage injection.

hafizrahman committed:
Refactor event trackings for better readability.

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift
	WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormEventLogger.swift
	WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormEventLoggerProtocol.swift
	WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductVariationFormEventLogger.swift

hafizrahman committed:
Re-delete code that were incorrectly added during trunk merge.

These were originally removed in this PR https://github.com/woocommerce/woocommerce-ios/pull/11248/files

hafizrahman committed:
Delete incorrectly added merge backup files.
